python/ray/tune/checkpoint_manager.py

```diff
@@ -198,6 +199,8 @@ def best_checkpoints(self):
     def _priority(self, checkpoint):
         result = flatten_dict(checkpoint.result)
         priority = result[self._checkpoint_score_attr]
+        if is_nan(priority):
+            priority = -float("inf")
         return -priority if self._checkpoint_score_desc else priority
```

Comment:
I think the priority can be a `tuple` that:

```suggestion
        if self._checkpoint_score_desc:
            priority = - priority
        return (not is_nan(priority), priority)
```

`(0, priority)` for `nan` and `(1, priority)` for others.

Reply:
Even further, we can have:

- `(0, iteration, priority)` for `nan`, the earlier checkpoint is worse.
- `(1, priority, iteration)` for not `nan` metrics.

Reply:
I was thinking about including iterations as well, but that would be a more substantial change in logic. @krfricke what do you think?

Reply:
Regarding iteration, I don’t think we have to treat nans as a special case. However we currently handle checkpoints with the same metric value, we can continue doing that even with nans.

Reply:
I do like the tulle approach though— makes it more explicit how nans are being treated.

Reply:
Let's go with that then!